flank should be optional in flank.yml

```
gcloud:
  app: ../outputs/apk/debug/app-debug.apk
  test: ../outputs/apk/androidTest/debug/app-debug-androidTest.apk
  project: flank-gradle
```

crashes doctor command because `flank:` is missing.

----

run command fails on:


```
gcloud:
  app: ../outputs/apk/debug/app-debug.apk
  test: ../outputs/apk/androidTest/debug/app-debug-androidTest.apk
  project: flank-gradle

flank:
```

Adding the `flank:` causes the following exception when running the real command, but the doctor safely passes.

```
Exception in thread "main" picocli.CommandLine$ExecutionException: Error while running command (ftl.cli.firebase.test.android.AndroidRunCommand@6a8658ff): com.fasterxml.jackson.module.kotlin.MissingKotlinParameterException: Instantiation of [simple type, class ftl.args.yml.FlankYml] value failed for JSON property flank due to missing (therefore NULL) value for creator parameter flank which is a non-nullable type
 at [Source: (StringReader); line: 6, column: 7] (through reference chain: ftl.args.yml.FlankYml["flank"])
	at picocli.CommandLine.execute(CommandLine.java:914)
	at picocli.CommandLine.access$700(CommandLine.java:104)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:1083)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:1051)
	at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:959)
	at picocli.CommandLine.parseWithHandlers(CommandLine.java:1242)
	at picocli.CommandLine.run(CommandLine.java:1701)
	at picocli.CommandLine.run(CommandLine.java:1647)
	at ftl.Main$Companion.main(Main.kt:38)
	at ftl.Main.main(Main.kt)
Caused by: com.fasterxml.jackson.module.kotlin.MissingKotlinParameterException: Instantiation of [simple type, class ftl.args.yml.FlankYml] value failed for JSON property flank due to missing (therefore NULL) value for creator parameter flank which is a non-nullable type
 at [Source: (StringReader); line: 6, column: 7] (through reference chain: ftl.args.yml.FlankYml["flank"])
	at com.fasterxml.jackson.module.kotlin.KotlinValueInstantiator.createFromObjectWith(KotlinValueInstantiator.kt:54)
	at com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:195)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:422)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1287)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4013)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3004)
	at ftl.args.AndroidArgs$Companion.load(AndroidArgs.kt:140)
	at ftl.args.AndroidArgs$Companion.load(AndroidArgs.kt:137)
	at ftl.cli.firebase.test.android.AndroidRunCommand.run(AndroidRunCommand.kt:27)
	at picocli.CommandLine.execute(CommandLine.java:906)
	... 9 more
```

--- a/test_runner/src/main/kotlin/ftl/args/yml/FlankYml.kt
+++ b/test_runner/src/main/kotlin/ftl/args/yml/FlankYml.kt
@@ -25,8 +25,12 @@ class FlankYmlParams(
 
 @JsonIgnoreProperties(ignoreUnknown = true)
 class FlankYml(
-    val flank: FlankYmlParams = FlankYmlParams()
+    // when an empty 'flank:' is present in a yaml then parsedFlank will be parsed as null.
+    @field:JsonProperty("flank")
+    private val parsedFlank: FlankYmlParams? = FlankYmlParams()
 ) {
+    val flank = parsedFlank ?: FlankYmlParams()
+
     companion object : IYmlMap {
         override val map = mapOf("flank" to FlankYmlParams.keys)
     }

--- a/test_runner/src/main/kotlin/ftl/args/yml/IosFlankYml.kt
+++ b/test_runner/src/main/kotlin/ftl/args/yml/IosFlankYml.kt
@@ -16,8 +16,11 @@ class IosFlankYmlParams(
 
 @JsonIgnoreProperties(ignoreUnknown = true)
 class IosFlankYml(
-    val flank: IosFlankYmlParams = IosFlankYmlParams()
+    // when an empty 'flank:' is present in a yaml then parsedFlank will be parsed as null.
+    @field:JsonProperty("flank")
+    private val parsedFlank: IosFlankYmlParams? = IosFlankYmlParams()
 ) {
+    val flank = parsedFlank ?: IosFlankYmlParams()
     companion object : IYmlMap {
         override val map = mapOf("flank" to IosFlankYmlParams.keys)
     }

--- a/test_runner/src/main/kotlin/ftl/doctor/Doctor.kt
+++ b/test_runner/src/main/kotlin/ftl/doctor/Doctor.kt
@@ -39,7 +39,7 @@ object Doctor {
 
         validArgs.forEach { (topLevelKey, keyList) ->
             val parsedKeys = mutableListOf<String>()
-            parsed[topLevelKey].fields().forEach { parsedKeys.add(it.key) }
+            parsed[topLevelKey]?.fields()?.forEach { parsedKeys.add(it.key) }
             val unknownKeys = parsedKeys - keyList
             if (unknownKeys.isNotEmpty()) result += "Unknown keys in $topLevelKey -> $unknownKeys\n"
         }

--- a/test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt
+++ b/test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt
@@ -1,5 +1,6 @@
 package ftl.args
 
+import com.google.common.truth.Truth.assertThat
 import ftl.config.Device
 import ftl.test.util.FlankTestRunner
 import ftl.test.util.TestHelper.assert
@@ -56,7 +57,7 @@ class AndroidArgsTest {
         expectedException.expect(RuntimeException::class.java)
         expectedException.expectMessage("Unsupported model id")
         AndroidArgs.load(
-            """
+                """
         gcloud:
           app: $appApk
           test: $testApk
@@ -72,7 +73,7 @@ class AndroidArgsTest {
         expectedException.expect(RuntimeException::class.java)
         expectedException.expectMessage("Unsupported version id")
         AndroidArgs.load(
-            """
+                """
         gcloud:
           app: $appApk
           test: $testApk
@@ -88,7 +89,7 @@ class AndroidArgsTest {
         expectedException.expect(RuntimeException::class.java)
         expectedException.expectMessage("Incompatible model")
         AndroidArgs.load(
-            """
+                """
         gcloud:
           app: $appApk
           test: $testApk
@@ -133,7 +134,7 @@ class AndroidArgsTest {
     fun androidArgsToString() {
         val androidArgs = AndroidArgs.load(androidNonDefault)
         assert(
-            androidArgs.toString(), """
+                androidArgs.toString(), """
 AndroidArgs
     gcloud:
       resultsBucket: mockBucket
@@ -163,7 +164,7 @@ AndroidArgs
     @Test
     fun androidArgsDefault() {
         val androidArgs = AndroidArgs.load(
-            """
+                """
         gcloud:
           app: $appApk
           test: $testApk
@@ -199,7 +200,7 @@ AndroidArgs
     @Test
     fun negativeOneTestShards() {
         val androidArgs = AndroidArgs.load(
-            """
+                """
         gcloud:
           app: $appApk
           test: $testErrorApk
@@ -214,5 +215,21 @@ AndroidArgs
             assert(testShardChunks.size, 2)
             testShardChunks.forEach { chunk -> assert(chunk.size, 1) }
         }
+    }
+
+    @Test
+    fun androidArgs_emptyFlank() {
+        val androidArgs = AndroidArgs.load(
+                """
+        gcloud:
+          app: $appApk
+          test: $testApk
+
+        flank:
+      """
+        )
+
+        assertThat(androidArgs.appApk).isEqualTo(appApk)
+        assertThat(androidArgs.testApk).isEqualTo(testApk)
     }
 }

--- a/test_runner/src/test/kotlin/ftl/args/IosArgsTest.kt
+++ b/test_runner/src/test/kotlin/ftl/args/IosArgsTest.kt
@@ -1,5 +1,6 @@
 package ftl.args
 
+import com.google.common.truth.Truth.assertThat
 import ftl.args.yml.FlankYml
 import ftl.args.yml.GcloudYml
 import ftl.args.yml.IosFlankYml
@@ -169,6 +170,24 @@ IosArgs
             assert(testShards, -1)
             assert(testShardChunks.size, 6)
             testShardChunks.forEach { chunk -> assert(chunk.size, 1) }
+        }
+    }
+
+    @Test
+    fun iosArgs_emptyFlank() {
+        val iosArgs = IosArgs.load(
+                """
+    gcloud:
+      test: $testPath
+      xctestrun-file: $xctestrunFile
+
+    flank:
+"""
+        )
+
+        with(iosArgs) {
+            assertThat(xctestrunZip).isEqualTo(testPath)
+            assertThat(xctestrunFile).isEqualTo(xctestrunFile)
         }
     }
 }

--- a/test_runner/src/test/kotlin/ftl/doctor/DoctorTest.kt
+++ b/test_runner/src/test/kotlin/ftl/doctor/DoctorTest.kt
@@ -67,6 +67,19 @@ Unknown keys in flank -> [three]
     }
 
     @Test
+    fun androidDoctorTest3() {
+        val lint = Doctor.validateYaml(
+                AndroidArgs, """
+gcloud:
+  app: .
+  test: .
+  project: .
+        """.trimIndent()
+        )
+        assertThat(lint).isEqualTo("")
+    }
+
+    @Test
     fun iosDoctorTest() {
         Doctor.checkIosCatalog()
         val lint = Doctor.validateYaml(IosArgs, Paths.get("src/test/kotlin/ftl/fixtures/flank.ios.yml"))
@@ -114,5 +127,18 @@ Unknown keys in flank -> [three]
 
 """.trimIndent()
         )
+    }
+
+    @Test
+    fun iosDoctorTest3() {
+        val lint = Doctor.validateYaml(
+                IosArgs, """
+gcloud:
+  project: .
+  test: .
+  xctestrun-file: .
+""".trimIndent()
+        )
+        assertThat(lint).isEqualTo("")
     }
 }

